### ISSUE TITLE
Question: Custom Service Provider Not Used Everywhere

### ISSUE THREAD

**Issue by rjperes** [dotnet/aspnetcore#58400]
So, I register my own IServiceProviderFactory<IServiceCollection> using WebApplicationBuilder.Host.UseServiceProviderFactory<IServiceCollection>, from which I return a custom IServiceProvider implementation. I see that it is called, and *appears* to work, however, when I try to inject some service into my controller, I see that it does not come from my IServiceProvider, but instead from an ServiceProviderEngineScope. Why is this? Shouldn't the registered factory be used for creating the global IServiceProvider?

**Comment by martincostello** [comment]
It's because each HTTP request creates a new scope using the root service provider, and then uses the service provider associated with that scope:

https://github.com/dotnet/aspnetcore/blob/049814ca468cad1ea1e29412e0aa3eea182a63c1/src/SignalR/common/Http.Connections/src/Internal/HttpConnectionDispatcher.cs#L764-L765

**Comment by rjperes** [comment]
But from where is IServiceScopeFactory obtained? If I registered my own IServiceProvider, I would expect it to come from there, but it isn't.

**Comment by martincostello** [comment]
As far as I can tell, it's implemented [as part of the built-in service provider](https://github.com/dotnet/runtime/blob/6221ddb3051463309801c9008f332b34361da798/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L54-L65) via [ServiceProviderEngineScope](https://github.com/dotnet/runtime/blob/6221ddb3051463309801c9008f332b34361da798/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs#L15). If your service provider implementation isn't being asked for the `IServiceScopeFactory`, then I don't know why.

**Comment by rjperes** [comment]
@martincostello, yes, that's what I'm saying: my custom service provider is not being asked for 
IServiceScopeFactory, so there must be another place where it is being constructed.

**Comment by rjperes** [comment]
This other service provider is used, for example, for the RequestServices property of the HttpContext.
